src/WheelsAndGoods.Application/Services/OrdersService.cs

```diff
@@ -99,5 +100,31 @@ public async Task DeleteOrder(Guid orderId)
                 order.IsDeleted = true;
             });
         }
+
+        public async Task<TakeOrderResponse> TakeOrder(Guid orderId, TakeOrderRequest request)
+        {
+            var order = await _unitOfWork.Orders.GetById(orderId, false);
+            var userId = Guid.Parse(_tokenReader.UserId);
+
+            if (order is null)
+            {
+                throw new NotFoundException("Order not found");
+            }
+            if (order.Customer.Id == userId || order.Status != Status.WaitingForContractor)
+            {
+                throw new AccessDeniedException("User has no access to take this order");
```

Comment:
It looks more like AppValidationException with different error messages, there is no problem with access, but there is a problem with violation of business (application) rules.